File: C++/Euler (C++)/src/problems/P001xP010/P003.cpp
#include <math.h>
#include <string>
#include <vector>

#include "../../util/basic.h"
#include "../../base/problem.h"
#include "../../util/prime.h"

/**
 * P003<br>
 * The prime factors of 13195 are 5, 7, 13 and 29.<br>
 * What is the largest prime factor of the number 600851475143 ?<br>
 * Ans: 6857
 */
class P003: public Problem
{
   public:
      P003()
      {
         problem_number = 3;
         variations = 2;
      }
      ~P003()
      {
      }

      std::string execute(int variation)
      {
         switch (variation)
         {
            case 0:
            case -1:
               return to_string(brute());
            case 1:
               return to_string(primeTest());
            default:
               return std::string("unavailable");
         }
         return NULL;
      }

   private:
      static const long n = 600851475143L;

      // factor the number, starting with 2's, then 3's, and on up. The last
      // factor is the largest.
      long brute()
      {
         return (basic::factors(n))->back();
      }

      int primeTest()
      {
         std::vector<int>* primes = prime::EratosthenesPrimesUpTo(sqrt(n));
         for (unsigned int ii = primes->size() - 1; ii >= 0; ii--)
         {
            if (n % primes->at(ii) == 0)
            {
            	int retVal = primes->at(ii);
            	delete primes;
               return retVal;
            }
         }
         delete primes;
         return -1;
      }

};

File: Java/Euler (Java)/src/problems/P001xP050/P003.java
package problems.P001xP050;

import java.util.List;

import util.Basic;
import util.Primes;
import base.Problem;

/**
 * P003<br>
 * The prime factors of 13195 are 5, 7, 13 and 29.<br>
 * What is the largest prime factor of the number 600851475143 ?<br>
 * Ans: 6857
 * @return
 */
public class P003 extends Problem
{
   public P003()
   {
      variations = 2;
   }

   private long n = 600851475143L;

   @Override
   protected String execute(int variation)
   {
      switch (variation)
      {
         case 0:
         case -1:
            return bruteForce() + "";
         case 1:
            return primeTest() + "";
      }
      return null;
   }

   /**
    * Brute force (mostly). Start with 2, then 3, then 5, and just factor it
    * 
    * @return
    */
   private long bruteForce()
   {
      List<Long> factors = Basic.getPrimeFactors(n);
      return factors.get(factors.size() - 1);
   }

   /**
    * Check to see if decreasing primes are the right answer
    * 
    * @return
    */
   private long primeTest()
   {
      List<Integer> primes = Primes.EratosthenesPrimesUpTo((int) Math.sqrt(n));
      for (int ii = primes.size() - 1; ii >= 0; ii--)
      {
         if (n % primes.get(ii) == 0)
            return primes.get(ii);
      }
      return -1;
   }

}
